cluster-autoscaler/processors/nodes/scale_down_set_processor.go

```diff
@@ -37,12 +39,65 @@ func NewCompositeScaleDownSetProcessor(orderedProcessorList []ScaleDownSetProces
 	}
 }
 
-// GetNodesToRemove selects nodes to remove.
-func (p *CompositeScaleDownSetProcessor) GetNodesToRemove(ctx *context.AutoscalingContext, candidates []simulator.NodeToBeRemoved, maxCount int) []simulator.NodeToBeRemoved {
-	for _, p := range p.orderedProcessorList {
-		candidates = p.GetNodesToRemove(ctx, candidates, maxCount)
+// FilterUnremovableNodes filters the passed removable candidates from unremovable nodes by calling orderedProcessorList in order
+func (p *CompositeScaleDownSetProcessor) FilterUnremovableNodes(ctx *context.AutoscalingContext, scaleDownCtx ScaleDownContext, candidates []simulator.NodeToBeRemoved) ([]simulator.NodeToBeRemoved, []simulator.UnremovableNode) {
+
+	unremovableNodes := []simulator.UnremovableNode{}
+	nodesToBeRemoved := []simulator.NodeToBeRemoved{}
+	nodesToBeRemoved = append(nodesToBeRemoved, candidates...)
+
+	for indx, p := range p.orderedProcessorList {
+		processorRemovableNodes, processorUnremovableNodes := p.FilterUnremovableNodes(ctx, scaleDownCtx, nodesToBeRemoved)
+
+		if !isValidCandidatesFiltering(candidates, processorRemovableNodes, processorUnremovableNodes) {
+			klog.Errorf("Scale deown set composite processor failed with processor at index %d", indx)
+		}
+
+		nodesToBeRemoved = processorRemovableNodes
+		unremovableNodes = append(unremovableNodes, processorUnremovableNodes...)
 	}
-	return candidates
+	return nodesToBeRemoved, unremovableNodes
+}
+
+// isValidCandidatesFeltering checks that exactly all candidates are splitted into removableNodes and unremovableNodes and loggign additional/missing nodes
+func isValidCandidatesFiltering(candidates []simulator.NodeToBeRemoved, removableNodes []simulator.NodeToBeRemoved, unremovableNodes []simulator.UnremovableNode) bool {
```

Comment:
This is a pretty costly check to run every loop. I wonder if it wouldn't make more sense as a unit test? All scale down set processors should pass it.

Reply:
well, @bskiba [suggested](https://github.com/kubernetes/autoscaler/pull/7307#discussion_r1786343013) to log additional/missing nodes, I don't have much preference but I am leaning towards skipping this costly check as it's also only for composite processor only so we can't say we guarantee this behaviour for all processors

Reply:
Yeah, good point. Feel free to remove, sorry for giving you the extra work here :/

Reply:
great, Reverted back to just check lengths, and we can count on unit tests to catch the cases of returning more or less nodes than candidates

Reply:
Should this function be removed?